Build failure on arm64 with GCC

**Describe the bug**
Since 10f5c38d2d3737464a3a09b2e73f3674641c760e I cannot build jxl on arm64 with GCC.
Clang compiles fine.
```
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h: In function 'void jxl::N_NEON::{anonymous}::FastTransposeBlock(const int16_t*, size_t, size_t, size_t, int16_t*, size_t)':
/libjxl/lib/jxl/fast_dct-inl.h:61:62: note: use '-flax-vector-conversions' to permit conversions between vectors with differing element types or numbers of subparts
   61 |       auto four0 = vtrnq_s32(vreinterpretq_s16_s32(a01.val[0]),
      |                                                              ^
/libjxl/lib/jxl/fast_dct-inl.h:61:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   61 |       auto four0 = vtrnq_s32(vreinterpretq_s16_s32(a01.val[0]),
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:62:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   62 |                              vreinterpretq_s16_s32(a23.val[0]));
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:63:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   63 |       auto four1 = vtrnq_s32(vreinterpretq_s16_s32(a01.val[1]),
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:64:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   64 |                              vreinterpretq_s16_s32(a23.val[1]));
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:74:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   74 |       auto four2 = vtrnq_s32(vreinterpretq_s16_s32(a45.val[0]),
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:75:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   75 |                              vreinterpretq_s16_s32(a67.val[0]));
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:76:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   76 |       auto four3 = vtrnq_s32(vreinterpretq_s16_s32(a45.val[1]),
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
In file included from /libjxl/lib/jxl/fast_dct.cc:13:
/libjxl/lib/jxl/fast_dct-inl.h:77:61: error: cannot convert 'int16x8_t' to 'int32x4_t'
   77 |                              vreinterpretq_s16_s32(a67.val[1]));
      |                                                    ~~~~~~~~~^
      |                                                             |
      |                                                             int16x8_t
In file included from /libjxl/third_party/highway/hwy/ops/arm_neon-inl.h:18,
                 from /libjxl/third_party/highway/hwy/highway.h:303,
                 from /libjxl/lib/jxl/fast_dct.cc:9:
/usr/lib/gcc/aarch64-alpine-linux-musl/10.3.1/include/arm_neon.h:5253:34: note:   initializing argument 1 of 'int16x8_t vreinterpretq_s16_s32(int32x4_t)'
 5253 | vreinterpretq_s16_s32 (int32x4_t __a)
      |                        ~~~~~~~~~~^~~
[ 72%] Building CXX object lib/CMakeFiles/jxl_dec-obj.dir/jxl/fields.cc.o
make[3]: *** [lib/CMakeFiles/jxl_dec-obj.dir/build.make:594: lib/CMakeFiles/jxl_dec-obj.dir/jxl/fast_dct.cc.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [CMakeFiles/Makefile2:3707: lib/CMakeFiles/jxl_dec-obj.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:5148: tools/CMakeFiles/cjxl.dir/rule] Error 2
make: *** [Makefile:2038: cjxl] Error 2
```

**To Reproduce**
Steps to reproduce the behavior via Docker on arm64:
```
from alpine:3.14

run apk add git cmake build-base libjpeg-turbo-dev libpng-dev zlib-dev
run git clone https://github.com/libjxl/libjxl.git
workdir /libjxl
run git submodule update --init --recursive
workdir /libjxl/build
run cmake ..
run make -j$(($(nproc) + 1)) cjxl djxl
```

**Environment**
 - OS: Alpine 3.14
 - Compiler version: GCC 10.3.1
 - CPU type: aarch64
```
Architecture:            aarch64
  CPU op-mode(s):        32-bit, 64-bit
  Byte Order:            Little Endian
CPU(s):                  4
  On-line CPU(s) list:   0-3
Vendor ID:               ARM
  Model name:            Neoverse-N1
    Model:               1
    Thread(s) per core:  1
    Core(s) per cluster: 4
    Socket(s):           -
    Cluster(s):          1
    Stepping:            r3p1
    BogoMIPS:            50.00
    Flags:               fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs
NUMA:
  NUMA node(s):          1
  NUMA node0 CPU(s):     0-3
Vulnerabilities:
  Itlb multihit:         Not affected
  L1tf:                  Not affected
  Mds:                   Not affected
  Meltdown:              Not affected
  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:            Mitigation; __user pointer sanitization
  Spectre v2:            Not affected
  Srbds:                 Not affected
  Tsx async abort:       Not affected
```

and the fix is not complete yet it seems: 

`cat  temp/build.log | grep error:`

```
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:95:48: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:96:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:97:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:98:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:99:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:100:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:101:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
/usr/aarch64-unknown-linux-gnu/tmp/portage/media-libs/libjxl-9999/work/libjxl-9999/lib/jxl/fast_dct-inl.h:102:54: error: cannot convert ‘__Int32x4_t’ to ‘int16x8_t’
```


full log with patch from https://github.com/libjxl/libjxl/pull/886 [libjxl-aarch64-incomplete-fix.zip](https://github.com/libjxl/libjxl/files/7580307/libjxl-aarch64-incomplete-fix.zip)

--- a/lib/jxl/fast_dct-inl.h
+++ b/lib/jxl/fast_dct-inl.h
@@ -58,10 +58,10 @@ HWY_NOINLINE void FastTransposeBlock(const int16_t* JXL_RESTRICT data_in,
       auto a01 = vtrnq_s16(a0, a1);
       auto a23 = vtrnq_s16(a2, a3);
 
-      auto four0 = vtrnq_s32(vreinterpretq_s16_s32(a01.val[0]),
-                             vreinterpretq_s16_s32(a23.val[0]));
-      auto four1 = vtrnq_s32(vreinterpretq_s16_s32(a01.val[1]),
-                             vreinterpretq_s16_s32(a23.val[1]));
+      auto four0 = vtrnq_s32(vreinterpretq_s32_s16(a01.val[0]),
+                             vreinterpretq_s32_s16(a23.val[0]));
+      auto four1 = vtrnq_s32(vreinterpretq_s32_s16(a01.val[1]),
+                             vreinterpretq_s32_s16(a23.val[1]));
 
       auto a4 = vld1q_s16(data_in + (i + 4) * stride_in + j);
       auto a5 = vld1q_s16(data_in + (i + 5) * stride_in + j);
@@ -71,10 +71,10 @@ HWY_NOINLINE void FastTransposeBlock(const int16_t* JXL_RESTRICT data_in,
       auto a45 = vtrnq_s16(a4, a5);
       auto a67 = vtrnq_s16(a6, a7);
 
-      auto four2 = vtrnq_s32(vreinterpretq_s16_s32(a45.val[0]),
-                             vreinterpretq_s16_s32(a67.val[0]));
-      auto four3 = vtrnq_s32(vreinterpretq_s16_s32(a45.val[1]),
-                             vreinterpretq_s16_s32(a67.val[1]));
+      auto four2 = vtrnq_s32(vreinterpretq_s32_s16(a45.val[0]),
+                             vreinterpretq_s32_s16(a67.val[0]));
+      auto four3 = vtrnq_s32(vreinterpretq_s32_s16(a45.val[1]),
+                             vreinterpretq_s32_s16(a67.val[1]));
 
       auto out0 =
           vcombine_s32(vget_low_s32(four0.val[0]), vget_low_s32(four2.val[0]));
@@ -92,14 +92,21 @@ HWY_NOINLINE void FastTransposeBlock(const int16_t* JXL_RESTRICT data_in,
                                vget_high_s32(four2.val[1]));
       auto out7 = vcombine_s32(vget_high_s32(four1.val[1]),
                                vget_high_s32(four3.val[1]));
-      vst1q_s16(data_out + j * stride_out + i, out0);
-      vst1q_s16(data_out + (j + 1) * stride_out + i, out1);
-      vst1q_s16(data_out + (j + 2) * stride_out + i, out2);
-      vst1q_s16(data_out + (j + 3) * stride_out + i, out3);
-      vst1q_s16(data_out + (j + 4) * stride_out + i, out4);
-      vst1q_s16(data_out + (j + 5) * stride_out + i, out5);
-      vst1q_s16(data_out + (j + 6) * stride_out + i, out6);
-      vst1q_s16(data_out + (j + 7) * stride_out + i, out7);
+      vst1q_s16(data_out + j * stride_out + i, vreinterpretq_s16_s32(out0));
+      vst1q_s16(data_out + (j + 1) * stride_out + i,
+                vreinterpretq_s16_s32(out1));
+      vst1q_s16(data_out + (j + 2) * stride_out + i,
+                vreinterpretq_s16_s32(out2));
+      vst1q_s16(data_out + (j + 3) * stride_out + i,
+                vreinterpretq_s16_s32(out3));
+      vst1q_s16(data_out + (j + 4) * stride_out + i,
+                vreinterpretq_s16_s32(out4));
+      vst1q_s16(data_out + (j + 5) * stride_out + i,
+                vreinterpretq_s16_s32(out5));
+      vst1q_s16(data_out + (j + 6) * stride_out + i,
+                vreinterpretq_s16_s32(out6));
+      vst1q_s16(data_out + (j + 7) * stride_out + i,
+                vreinterpretq_s16_s32(out7));
     }
   }
 }

